src/clipboard.c

```diff
@@ -0,0 +1,538 @@
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+  else if (!remote_desktop_session_is_clipboard_enabled (
+               (RemoteDesktopSession *)session))
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Clipboard not enabled");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  xdp_dbus_impl_clipboard_call_selection_read (impl,
+                                               arg_session_handle,
+                                               arg_mime_type,
+                                               NULL,
+                                               NULL,
+                                               selection_read_done,
+                                               g_object_ref (invocation));
+
+  return G_DBUS_METHOD_INVOCATION_HANDLED;
+}
+
+static void
+clipboard_iface_init (XdpDbusClipboardIface *iface)
+{
+  iface->handle_request_clipboard = handle_request_clipboard;
+
+  iface->handle_selection_read = handle_selection_read;
+  iface->handle_selection_write = handle_selection_write;
+  iface->handle_set_selection = handle_set_selection;
+  iface->handle_selection_write_done = handle_selection_write_done;
+}
+
+static void
+clipboard_init (Clipboard *clipboard)
+{
+  xdp_dbus_clipboard_set_version (XDP_DBUS_CLIPBOARD (clipboard), 1);
+}
+
+static void
+clipboard_class_init (ClipboardClass *klass)
+{
+}
+
+static void
+selection_transfer_cb (XdpDbusImplClipboard *impl,
+                       const char *arg_session_handle,
+                       const char *arg_mime_type,
+                       guint arg_serial,
+                       gpointer data)
+{
+  GDBusConnection *connection =
+      g_dbus_proxy_get_connection (G_DBUS_PROXY (impl));
+  g_autoptr (Session) session = lookup_session (arg_session_handle);
+
+  RemoteDesktopSession *remote_desktop_session =
+      (RemoteDesktopSession *)session;
+
+  if (remote_desktop_session &&
+      remote_desktop_session_is_clipboard_enabled (remote_desktop_session) &&
+      !session->closed)
+    {
+      g_dbus_connection_emit_signal (
+          connection,
+          session->sender,
+          "/org/freedesktop/portal/desktop",
+          "org.freedesktop.portal.Clipboard",
+          "SelectionTransfer",
+          g_variant_new (
+              "(osu)", arg_session_handle, arg_mime_type, arg_serial),
+          NULL);
+    }
+}
+
+static void
+selection_owner_changed_cb (XdpDbusImplClipboard *impl,
+                            const char *arg_session_handle,
+                            GVariant *arg_options,
+                            gpointer data)
+{
+  GDBusConnection *connection =
+      g_dbus_proxy_get_connection (G_DBUS_PROXY (impl));
+  g_autoptr (Session) session = lookup_session (arg_session_handle);
```

Comment:
Might want to move the assignment out of the declaration, and then NULL-check it.

The signal we're handling is from the `impl` but the session is local and might have been destroyed already.